### DISCUSSION TITLE
When do nodes have None position?

### DISCUSSION THREAD

**Discussion by mightyiam** [wooorm/markdown-rs#97] [Q&A] [answered]
When a tree has been parsed from a string, will they always have positions?

Edit: I peeked at the construction of one node: Html and found that there is a case where position is None. But there are many nodes. I hope there's a straightforward answer to this. I'd also be able to pass on that option to my users, if it can be None in my usage, but I'd better be certain of it.

To be clear, my usage is:
1. parse markdown using to_mdast with default options.
2. Iterate over the children.

Can I expect all of these children to have Some positions?

**Comment by ChristianMurphy** [comment]
Welcome @mightyiam! 👋 

Taking a step back, this feels a lot like an [XY Problem](https://meta.stackexchange.com/a/66378).
What are you trying to do with positional information? What concerns do you have with `None` positional info?

> When a tree has been parsed from a string, will they always have positions?

No, they will not.
There are referential constructs like: HTML, footnotes, and reference links which exist in multiple places or get synthesized and can have `None` positional information.
Also when plugin support is added (https://github.com/wooorm/markdown-rs/issues/32), plugins can add nodes which didn't exist in the original document, and can have `None` position.

**Reply by mightyiam** [reply]
Here's a snippet from the readme of an open source project I'm cooking:

````md

<!-- range description -->
Upsert Markdown range delineated by HTML comments.
<!-- range-end -->

<!-- range usage -->
Example input document:

```md
The markdown ranger went to town.

<!-- range what-happened -->
The markdown ranger met a clown.
<!-- range-end -->

The markdown ranger did not frown.
```

The range between the comments can be upserted or checked.

Id must be unique.

Files must be valid UTF-8.
<!-- range-end -->

See also CLI [mdranger](https://crates.io/crates/mdranger).

````

Here's the primary public function:

```rust
pub fn upsert_document(
    document: &mut String,
    target_range_id: &Id,
    upsertion: &str,
) -> Result<Report, SearchError> {
    let target_range = search(document, target_range_id)?;
    let report = with_range::upsert(document, &target_range, upsertion);
    Ok(report)
}
```

The error type looks like this:

```rust

#[derive(Debug, thiserror::Error, Clone, PartialEq, Eq)]
pub enum Error {
    #[error("start comment where comment that preceeds it is start comment: {0:?}, {1:?}")]
    UnexpectedStart(Position, Position),
    #[error("end comment where either no range comment preceeds it or end comment does: {0:?}")]
    UnexpectedEnd(Position),
    #[error("start comment not followed by end comment: {0:?}")]
    Unclosed(Position),
    #[error("invalid range id: {1} at: {0:?}")]
    InvalidId(Position, IdError),
    #[error("start comment with same id as another: {0:?}, {1:?}")]
    DuplicateId(Position, Position),
    #[error("range with provided id not found")]
    NotFound,
}
```

Those Position structs are not directly from the markdown crate, but have the same shape and same goes for the Point structs.

Should all my Position fields be options, then?

**Reply by mightyiam** [reply]
All these positions are for HTML comments. Likely doesn't matter that they're comments, because all markdown seems to know is that they're HTML. From my experience so far, they all had positions.

**Reply by wooorm** [reply]
I recommend to always handle `None` positions. Nodes that were not in the original document, but generated by plugins, will not have positions.

**Reply by mightyiam** [reply]
Sorry. I don't see how those will exist in my usage. I don't allow my users to set up any plugins and I don't do so either. All my users provide is a document string.

**Reply by ChristianMurphy** [reply]
@mightyiam let me simplify this.
Position is always optional, period.
It will come through most of the time, but don't bake the assumption it will always be there into your code, it is incorrect, it can and will break.

If you choose bypass the SemVer API and make assumptions about the implementation, see my comments above on when the current implementation, as of alpha 15, will likely return `None`

**Reply by mightyiam** [reply]
Moving my comment here: Sorry. I've not been as clear as I could have. What I would like to be able to unwrap the position of are lines that start with `<!-- `. Is there any chance I'll get a different answer for that, such as "yes, these will always be Html nodes with Some(position)"?

And @ChristianMurphy's response was:

> To reiterate at a standard/API level, no.
> `markdown-rs` nodes are based on unist and mdast.
> Position is never guaranteed. https://github.com/syntax-tree/unist#node
>
>In the current implementation, alpha 15, you will most likely get a position for HTML nodes.
>
> However that is not an API guarantee, it may break.
> Rust focuses on correctness and guarantees, the safest implementation is assume `None` can happen.
>
> That said if you find places where position is missing and you believe it could be provided happy to discuss implementation improvements.

**Reply by mightyiam** [reply]
Thank you. I understand you don't want to make this guarantee. It's just that my entire implementation relies on these positions. Perhaps I should replace the parser.

**Reply by ChristianMurphy** [reply]
Completely up to you.
Part of my question to what you wanted to do was to understand your end goal.
If you want to know what's between two comments, using the AST, rather than position, is far more robust.

There are example of this in the JS implementation of mdast utils.

https://github.com/syntax-tree/mdast-zone

https://github.com/syntax-tree/mdast-util-heading-range

**Reply by mightyiam** [reply]
What I do is upsert a range delineated by two comments. Therefore I need to know the range between those comments. markdown-rs seems to work great for this purpose, but its authors are unwilling to guarantee positions.

**Reply by ChristianMurphy** [reply]
This still feels like an XY Problem.

Why not modify the AST and serialize the AST back to HTML or Markdown?
It sound like you want to modify structure, that's what syntax trees are for, you can modify them safely without mucking with positions.

**Reply by mightyiam** [reply]
I appreciate the advice. Yes, that would have been great, but if I do that, I would end up reformatting the entire document, which must not occur.

**Reply by ChristianMurphy** [reply, marked as answer]
> I would end up reformatting the entire document, which must not occur.

Then an Abstract Syntax Tree (AST) based project isn't suitable.
Abstract syntax trees focus on the abstract structure rather than syntax detail.
Positions in most (likely all) AST based libraries are designed for debugging and source mapping, not editing.

Most of the major rust based markdown parsers are AST based including:
* https://github.com/wooorm/markdown-rs
* https://github.com/kivikakk/comrak
* https://github.com/raphlinus/pulldown-cmark

What you are describing @mightyiam sounds more like a _Concrete_ Syntax Tree (CST), which preserves all detail about the text so it can be exactly reconstructed.
There are very few libraries out there https://tree-sitter.github.io/tree-sitter is the only one I'm currently aware of that has a CST for Markdown along with Rust bindings.

**Comment by mightyiam** [comment]


**Reply by ChristianMurphy** [reply]
To reiterate at a standard/API level, no.
`markdown-rs` nodes are based on unist and mdast.
Position is **never** guaranteed. https://github.com/syntax-tree/unist#node

In the current implementation, alpha 15, you will most likely get a position for HTML nodes.

However that is not an API guarantee, it may break.
Rust focuses on correctness and guarantees, the safest implementation is assume `None` can happen.

That said if you find places where position is missing and you believe it could be provided happy to discuss implementation improvements.

**Reply by mightyiam** [reply]
Sorry, I commented as an answer unintentionally. I deleted it and reposted it in your answer. Perhaps you could delete, as well.
